File: transpile/giada/pizzapazza/math/Z4Whirlpool.js
/**
 * The whirlpool
 *
 * @author gianpiero.diblasi
 */
class Z4Whirlpool {

   type = 0;

   angle = new Z4FancifulValue().setConstant(new Z4SignedValue().setValue(45).setSign(Z4Sign.POSITIVE)).setRandom(Z4SignedRandomValue.classic(0).setSign(Z4Sign.POSITIVE));

  constructor(type) {
    this.type = type;
  }

  /**
   * Sets the angle
   *
   * @param angle The angle
   * @return This Z4Whirlpool
   */
   setAngle(angle) {
    this.angle = angle;
    return this;
  }

  /**
   * Returns the angle
   *
   * @return The angle
   */
   getAngle() {
    return this.angle;
  }

  /**
   * Returns a Z4Whirlpool without whirlpool
   *
   * @return The Z4Whirlpool
   */
  static  none() {
    return new Z4Whirlpool(0);
  }

  /**
   * Returns a Z4Whirlpool with forward whirlpool
   *
   * @return The Z4Whirlpool
   */
  static  forward() {
    return new Z4Whirlpool(1);
  }

  /**
   * Returns a Z4Whirlpool with backward whirlpool
   *
   * @return The Z4Whirlpool
   */
  static  backward() {
    return new Z4Whirlpool(2);
  }

  /**
   * Returns if this Z4Whirlpool has no whirlpool
   *
   * @return true if this Z4Whirlpool has no whirlpool
   */
   isNone() {
    return this.type === 0;
  }

  /**
   * Returns if this Z4Whirlpool has a forward whirlpool
   *
   * @return true if this Z4Whirlpool has a forward whirlpool
   */
   isForward() {
    return this.type === 1;
  }

  /**
   * Returns if this Z4Whirlpool has a backward whirlpool
   *
   * @return true if this Z4Whirlpool has a backward whirlpool
   */
   isBackward() {
    return this.type === 2;
  }
}

File: src/pizzapazza/math/Z4Whirlpool.java
package pizzapazza.math;

import pizzapazza.util.Z4JSONable;
import simulation.js.$Object;

/**
 * The whirlpool
 *
 * @author gianpiero.diblasi
 */
public class Z4Whirlpool implements Z4JSONable {

  private final Z4WhirlpoolBehavior behavior;
  private final Z4FancifulValue angle;

  /**
   * Creates the object
   *
   * @param behavior The behavior
   * @param angle The angle
   */
  public Z4Whirlpool(Z4WhirlpoolBehavior behavior, Z4FancifulValue angle) {
    super();
    this.behavior = behavior;
    this.angle = angle;
  }

  /**
   * Returns the behavior
   *
   * @return The behavior
   */
  public Z4WhirlpoolBehavior getWhirlpoolBehavior() {
    return this.behavior;
  }

  /**
   * Returns the angle
   *
   * @return The angle
   */
  public Z4FancifulValue getAngle() {
    return this.angle;
  }

  @Override
  public $Object toJSON() {
    $Object json = new $Object();
    json.$set("behavior", this.behavior);
    json.$set("angle", this.angle.toJSON());
    return json;
  }

  /**
   * Creates a Z4Whirlpool from a JSON object
   *
   * @param json The JSON object
   * @return the whirlpool
   */
  public static Z4Whirlpool fromJSON($Object json) {
    return new Z4Whirlpool(json.$get("behavior"), Z4FancifulValue.fromJSON(json.$get("angle")));
  }
}
